File: modules/wem-webapp/src/main/webapp/admin/common/js/data/Value.ts
module api.data {

    export class Value implements api.Equitable {

        private value: Object;

        private type: ValueType;

        constructor(value: Object, type: ValueType) {
            api.util.assertNotNull(value, "value of a Value cannot be null");
            api.util.assertNotNull(type, "type of a Value cannot be null");
            this.value = value;
            this.type = type;
        }

        asObject(): Object {
            return this.value;
        }

        asString(): string {

            return this.type.valueToString(this);
        }

        isRootDataSet(): boolean {
            return ValueTypes.DATA.toString() == this.type.toString();
        }

        asRootDataSet(): RootDataSet {
            api.util.assert(this.value instanceof RootDataSet, "Expected value to be a RootDataSet: " + api.util.getClassName(this.value));
            return <RootDataSet>this.value;
        }

        getType(): ValueType {
            return this.type;
        }

        setValue(value: Object) {
            this.value = value;
        }

        equals(o: api.Equitable): boolean {

            if (!(o instanceof Value)) {
                return false;
            }

            var other = <Value>o;

            if (!api.EquitableHelper.objectEquals(this.value, other.value)) {
                return false;
            }

            if (!api.EquitableHelper.equals(this.type, other.type)) {
                return false;
            }

            return true;
        }
    }
}

File: modules/wem-api/src/main/java/com/enonic/wem/api/value/Value.java
package com.enonic.wem.api.value;

public abstract class Value<T>
{
    protected final T object;

    protected final ValueType type;

    public Value( final ValueType type, final T object )
    {
        this.type = type;
        this.object = object;

        if ( this.object == null )
        {
            throw new ValueException( "Null value not allowed for [%s]", this.type );
        }
    }

    public final T getObject()
    {
        return this.object;
    }

    public final ValueType getType()
    {
        return this.type;
    }

    public String asString()
    {
        return this.object.toString();
    }

    @Override
    public final String toString()
    {
        return asString();
    }
}
